snuba/query/expressions.py

```diff
@@ -0,0 +1,215 @@
+
+    def __iter__(self) -> Iterator[Expression]:
+        """
+        Traverse the subtree in a postfix order.
+        The order here is arbitrary, postfix is chosen to follow the same
+        order we have in the transform method.
+        """
+        for child in self._get_children():
+            for sub in child:
+                yield sub
+        yield self
+
+    @abstractmethod
+    def _get_children(self) -> Sequence[Expression]:
+        """
+        To be implemented by the subclasses to provide the list of children
+        for iteration/transformation.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def _duplicate_with_new_children(self, children: Sequence[Expression]) -> Expression:
+        """
+        Return a new instance of the expression with a new set of children.
+        """
+        raise NotImplementedError
+
+
+@dataclass(frozen=True)
+class Literal(Expression):
+    """
+    A literal in the SQL expression
+    """
+    value: str
+
+    def format(self) -> str:
+        raise NotImplementedError
+
+    def transform(self, func: Callable[[Expression], Expression]) -> Expression:
+        return func(self)
+
+    def __iter__(self) -> Iterator[Expression]:
+        yield self
+
+
+@dataclass(frozen=True)
+class Column(Expression):
+    """
+    Represent a column in the schema of the dataset.
+    """
+    column_name: str
+    table_name: Optional[str]
+
+    def format(self) -> str:
+        raise NotImplementedError
+
+    def transform(self, func: Callable[[Expression], Expression]) -> Expression:
+        return func(self)
+
+    def __iter__(self) -> Iterator[Expression]:
+        yield self
+
+
+@dataclass(frozen=True)
+class FunctionCall(HierarchicalExpression):
+    """
+    Represents an expression that resolves to a function call on Clickhouse.
+    This class also represent conditions. Since Clickhouse supports both the conventional
+    infix notation for condition and the functional one, we converge into one
+    representation only in the AST to make query processing easier.
+    A query processor would not have to care of processing both functional conditions
+    and infix conditions.
+    """
+    function_name: str
+    parameters: Sequence[Expression]
+
+    def format(self) -> str:
+        raise NotImplementedError
+
+    def _get_children(self) -> Sequence[Expression]:
+        return self.parameters
+
+    def _duplicate_with_new_children(self, children: Sequence[Expression]) -> Expression:
+        return FunctionCall(self.alias, self.function_name, children)
+
+
+@dataclass(frozen=True)
+class Aggregation(HierarchicalExpression):
+    """
+    Represents an aggregation function to be applied to an expression in the
+    current query.
+
+    TODO: I don't think this should exist, but as of now a lot of our query
+    processing still relies on aggregation being a first class concept.
+    """
+    function_name: str
+    parameters: Sequence[Expression]
```

Comment:
Do you have any initial thoughts on how this will support parametric functions (e.g. `topK(5)(col)`)?

Reply:
It seems it should be a subclass of FunctionCall and I was planning to address that when needed to reduce the size here. But I can add it now if you prefer.

Reply:
I'll leave that up to you as long as there's a plan.